components/public-api/gitpod/v1/workspaces.proto

```diff
@@ -0,0 +1,328 @@
+    rpc ListWorkspaces(ListWorkspacesRequest) returns (ListWorkspacesResponse) {}
+
+    // GetWorkspace returns a single workspace.
+    rpc GetWorkspace(GetWorkspaceRequest) returns (GetWorkspaceResponse) {}
+
+    // CreateAndStartWorkspace creates a new workspace and starts it.
+    rpc CreateAndStartWorkspace(CreateAndStartWorkspaceRequest) returns (CreateAndStartWorkspaceResponse) {}
+
+    // StartWorkspace starts an existing workspace.
+    rpc StartWorkspace(StartWorkspaceRequest) returns (StartWorkspaceResponse) {}
+
+    // GetRunningWorkspaceInstance returns the currently active instance of a workspace.
+    // Errors:
+    //   FAILED_PRECONDITION: if a workspace does not a currently active instance
+    //
+    rpc GetActiveWorkspaceInstance(GetActiveWorkspaceInstanceRequest) returns (GetActiveWorkspaceInstanceResponse) {}
+
+    // GetWorkspaceInstanceOwnerToken returns the owner token of a workspace instance.
+    // Note: the owner token is not part of the workspace instance status so that we can scope its access on the
+    //       API function level.
+    rpc GetWorkspaceInstanceOwnerToken(GetWorkspaceInstanceOwnerTokenRequest) returns (GetWorkspaceInstanceOwnerTokenResponse) {}
+
+    // ListenToWorkspaceInstance listens to workspace instance updates.
+    rpc ListenToWorkspaceInstance(ListenToWorkspaceInstanceRequest) returns (stream ListenToWorkspaceInstanceResponse) {}
+
+    // ListenToImageBuildLogs streams (currently or previously) running workspace image build logs
+    rpc ListenToImageBuildLogs(ListenToImageBuildLogsRequest) returns (stream ListenToImageBuildLogsResponse) {}
+
+    // StopWorkspace stops a running workspace (instance).
+    // Errors:
+    //   NOT_FOUND:           the workspace_id is unkown
+    //   FAILED_PRECONDITION: if there's no running instance
+    rpc StopWorkspace(StopWorkspaceRequest) returns (stream StopWorkspaceResponse) {}
+
+}
+
+message ListWorkspacesRequest {
+    Pagination pagination = 1;
+    google.protobuf.FieldMask field_mask = 2;
+}
+message ListWorkspacesResponse {
+    message WorkspaceAndInstance {
+        Workspace result = 1;
+        WorkspaceInstance last_active_instances = 2;
+    }
+
+    google.rpc.Status response_status = 1;
+
+    string next_page_token = 2;
+
+    repeated WorkspaceAndInstance result = 3;
+}
+
+message GetWorkspaceRequest {
+    string workspace_id = 1;
+}
+message GetWorkspaceResponse {
+    google.rpc.Status response_status = 1;
+
+    Workspace result = 2;
+}
+
+message CreateAndStartWorkspaceRequest {
+    string idempotency_token = 1;
+
+    string context_url = 2;
+
+    oneof prebuild {
+        // if true, and there's a prebuild available (not running), use that prebuild
+        bool if_available = 3;
+        // uses a particular prebuild - the prebuild ID is scoped by the context URL.
+        // Use the PrebuildService to get ahold of the prebuild_id.
+        string prebuild_id = 4;
```

Comment:
Why would I expicitly pass a prebuild_id?

Reply:
Today's `createWorkspace` call is rather complicated because it mixes prebuild selection and workspace creation to provide a form of "transactional behaviour". By making prebuilds an explicit conern this becomes a lot easier, and we can factor prebuild selection out of workspace creation.

Reply:
In the original design the prebuild_id would be matched against the context URL (basically scoped by the context URL). That feels very complicated. 

In the most recent draft the "source" of the workspace is a oneof `context_url` or `prebuild_id`.

Reply:
_I think the Workspace APIs should NOT accept a prebuild ID_. Here's why:

The API shouldn't necessarily encode how things work today, but much more the experience we want a user to have. We can use the "time to first 200" as a metric for the API UX.

With this in mind, we should eliminate ambiguity and concepts from the API calls which are not needed, or which the system can compute/discover for itself. In this case, I believe the system should be able to lookup if a prebuild for a context exists, and use it when it does - without the explicitly defining the prebuild ID. 

As a user, it's easier for me to think of workspaces in terms of contexts, and not prebuilds. This means that it will be easier for me to only supply the context and not the prebuild id when I'm integrating. This in turn reflects badly on us because the user is not getting the benefits of prebuilds.